### PR TITLE
fix: allow single quotes in key=value pairs

### PR DESCRIPTION
**Pull request by RodEsp**
### What does this PR do?
Allows using `'` in the value  of name-value pairs for flags such as `force:data:record:create`'s `--values`.

### What issues does this PR fix or reference?
https://github.com/forcedotcom/cli/issues/1820
@W-12105706@

### CODE REVIEW

File: src/dataCommand.ts [hunk 1]
```diff
@@ -222,48 +222,26 @@ export abstract class DataCommand extends SfdxCommand {
...
-          currentToken = [];
-        }
-      } else {
-        currentToken.push(currentChar);
-      }
-    }
+    // The regex below uses a positive lookbehind and a positive lookahead to split the string into key=value pairs
+    //   regardless of whitespace or single quotes in the values
+    // Use regex101.com and paste this regex to see how it works
+    const matches = text.match(/(\S+)=(?<=['"]?)([\d\w'"\s]+)(?=\s|$)/g);
```
Comment (peternhale) [inline src/dataCommand.ts:239]: @RodEsp the `\d` in `[\d\w'"\s]` is redundant, numbers are covered by `\w`

Comment (peternhale) [reply src/dataCommand.ts:239]: I love all the red in the diff here :)

**Review by peternhale — CHANGES_REQUESTED** [review]
I know there are not a lot UTs in this modules, but maybe one or two more around this change.

File: src/dataCommand.ts (hunk 1 above)
Comment (WillieRuemmele) [reply src/dataCommand.ts:239]: I tried pasting into regex101, and it had errors?

you can save your regex101 like https://regex101.com/r/gYqdkb/1 (ctrl+s) might be nice to build out an example there and save the url in the comments

Comment (peternhale) [reply src/dataCommand.ts:239]: @WillieRuemmele you need to pick the regex flavor. I chose ECMAScript and it worked for me.

**Comment by RodEsp** [comment]
Heya @peternhale, this is ready for another look. 

Unfortunately, I could not get the regex to work with both weird Unicode characters and values like `key="leftvalue rightvalue"` at the same time, so it looks like we're stuck with the current lengthy code that checks for the existence of `'`s and `"`s.

File: src/dataCommand.ts [hunk 2]
```diff
@@ -238,13 +238,17 @@ export abstract class DataCommand extends SfdxCommand {
     const keyValuePairs: string[] = [];
 
     const trimmedText = text.trim();
+
+    const singleQuoteCount = (trimmedText.match(/'/g) || []).length;
+    const doubleQuoteCount = (trimmedText.match(/"/g) || []).length;
+
     for (const currentChar of trimmedText) {
-      const isSeparator = separator.exec(currentChar);
+      const isSeparator = separator.exec(currentChar) !== null;
```
Comment (peternhale) [inline src/dataCommand.ts:246]:
```suggestion
      const isSeparator = separator.test(currentChar);
```

**Review by peternhale — CHANGES_REQUESTED** [review]
@RodEsp I found another valid string that does not make it through the function intact, but it probably would of failed in the current code too.
```
  it('should allow double and single quotes in key=value pairs', () => {
    const dict = dataCommand.testStringToDictionary("key='39° 44' 31.3548\" N'");

    expect(dict).to.deep.equal({
      key: '39° 44\' 31.3548" N',
    });
  });
```

**Comment by RodEsp** [comment]
Yeah, that didn't work without the changes in this PR either. The problem is with values that have quoted strings inside quoted strings, e.g. `key="this is a 'quote'"` or `key='this is a "quote"'`. I think it'll be a greater effort to get that to work, but for the sake of fixing https://github.com/forcedotcom/cli/issues/1820 asap I'd like to merge this now and I created https://github.com/forcedotcom/cli/issues/1837.

**Review by peternhale — APPROVED** [review]

**Pull request merged**